packages/template-retail-react-app/app/partials/product-view/index.jsx

```diff
@@ -76,161 +76,340 @@ const ButtonWithRegistration = withRegistration(Button)
-    const {
-        isOpen: isAddToCartModalOpen,
-        onOpen: onAddToCartModalOpen,
-        onClose: onAddToCartModalClose
-    } = useAddToCartModalContext()
-    const theme = useTheme()
-    const [showOptionsMessage, toggleShowOptionsMessage] = useState(false)
-    const {
-        showLoading,
-        showInventoryMessage,
-        inventoryMessage,
-        quantity,
-        minOrderQuantity,
-        setQuantity,
-        variant,
-        variationParams,
-        variationAttributes,
-        stockLevel,
-        stepQuantity
-    } = useProduct(product, isProductPartOfSet)
-    const canAddToWishlist = !isProductLoading
-    const canOrder =
-        !isProductLoading &&
-        variant?.orderable &&
-        parseInt(quantity) > 0 &&
-        parseInt(quantity) <= stockLevel
+const ProductView = forwardRef(
+    (
+        {
+            product,
+            category,
+            showFullLink = false,
+            imageSize = 'md',
+            isWishlistLoading = false,
+            addToCart,
+            updateCart,
+            addToWishlist,
+            updateWishlist,
+            isProductLoading,
+            isProductPartOfSet = false,
+            variantSelected = () => {},
+            validate = (variant, quantity, stockLevel) =>
+                variant?.orderable && quantity > 0 && quantity <= stockLevel
+        },
+        ref
+    ) => {
+        const intl = useIntl()
+        const history = useHistory()
+        const location = useLocation()
+        const {
+            isOpen: isAddToCartModalOpen,
+            onOpen: onAddToCartModalOpen,
+            onClose: onAddToCartModalClose
+        } = useAddToCartModalContext()
+        const theme = useTheme()
+        const [showOptionsMessage, toggleShowOptionsMessage] = useState(false)
+        const {
+            showLoading,
+            showInventoryMessage,
+            inventoryMessage,
+            quantity,
+            minOrderQuantity,
+            setQuantity,
+            variant,
+            variationParams,
+            variationAttributes,
+            stockLevel,
+            stepQuantity
+        } = useProduct(product, isProductPartOfSet)
+        const canAddToWishlist = !isProductLoading
+        const isProductASet = product?.type.set
 
-    const isProductASet = product?.type.set
+        const renderActionButtons = () => {
+            const buttons = []
 
-    const renderActionButtons = () => {
-        const buttons = []
+            const handleCartItem = async () => {
+                // Validate that all attributes are selected before proceeding.
+                const hasValidSelection = validate(variant, quantity, stockLevel)
 
-        const handleCartItem = async () => {
-            if (!canOrder) {
-                toggleShowOptionsMessage(true)
-                return null
+                if (!hasValidSelection) {
+                    !isProductASet && toggleShowOptionsMessage(true)
```

Comment:
```suggestion
                    if (!isProductASet) toggleShowOptionsMessage(true)
```
Y'ain't inside a JSX component here, use regular flow controls!

Reply:
👍